Sources/NIO/BaseSocketChannel.swift

```diff
@@ -14,6 +14,190 @@
 
 import NIOConcurrencyHelpers
 
+struct SocketChannelLifecycleManager {
+    // MARK: Types
+    enum State {
+        case neverRegistered
+        case registeredNeverActivated
+        case registered
+        case activated
+        case closed
+    }
+
+    enum Event {
+        case activate
+        case deactivate
+        case register
+        case close
+    }
+
+    // MARK: properties
+    // this is queried from the Channel, ie. must be thread-safe
+    internal let isActiveAtomic: Atomic<Bool>
+    // these are only to be accessed on the EventLoop
+    internal let channelPipeline: ChannelPipeline
+    private let eventLoop: EventLoop
+    private var currentState: State = .neverRegistered {
+        didSet {
+            assert(self.eventLoop.inEventLoop)
+            switch (oldValue, self.currentState) {
+            case (_, .activated):
+                self.isActiveAtomic.store(true)
+            case (.activated, _):
+                self.isActiveAtomic.store(false)
+            default:
+                ()
+            }
+        }
+    }
+
+    // MARK: API
+    internal init(eventLoop: EventLoop,
+                channelPipeline: ChannelPipeline,
+                isActiveAtomic: Atomic<Bool>) {
+        self.isActiveAtomic = isActiveAtomic
+        self.channelPipeline = channelPipeline
+        self.eventLoop = eventLoop
+    }
+
+    // this is called from Channel's deinit, so don't assert we're on the EventLoop!
+    internal var canBeDestroyed: Bool {
+        if case .closed = self.currentState {
+            return true
+        } else {
+            return false
+        }
+    }
+
+    @inline(__always) // so we're not actually returning a closure
+    public mutating func moveState(event: Event, promise: EventLoopPromise<()>?) -> (() -> Void) {
+        assert(self.eventLoop.inEventLoop)
+
+        switch (self.currentState, event) {
+        // origin: .neverRegistered
+        case (.neverRegistered, .activate):
```

Comment:
Do we intend to allow activating an unregistered socket? I think I'd kinda prefer we didn't.